syntax = "proto3";

package user;

option go_package = "github.com/noorshaik95/axum-grafana-example/services/user-auth-service/api/proto";

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

// User service for CRUD operations and authentication
service UserService {
  // Authentication
  rpc Login(LoginRequest) returns (LoginResponse);
  rpc Register(RegisterRequest) returns (RegisterResponse);
  rpc RefreshToken(RefreshTokenRequest) returns (RefreshTokenResponse);
  rpc ValidateToken(ValidateTokenRequest) returns (ValidateTokenResponse);
  rpc Logout(LogoutRequest) returns (google.protobuf.Empty);

  // User CRUD
  rpc CreateUser(CreateUserRequest) returns (UserResponse);
  rpc GetUser(GetUserRequest) returns (UserResponse);
  rpc UpdateUser(UpdateUserRequest) returns (UserResponse);
  rpc DeleteUser(DeleteUserRequest) returns (google.protobuf.Empty);
  rpc ListUsers(ListUsersRequest) returns (ListUsersResponse);

  // Profile Management
  rpc GetProfile(GetProfileRequest) returns (ProfileResponse);
  rpc UpdateProfile(UpdateProfileRequest) returns (ProfileResponse);
  rpc ChangePassword(ChangePasswordRequest) returns (google.protobuf.Empty);

  // Role Management (RBAC)
  rpc AssignRole(AssignRoleRequest) returns (google.protobuf.Empty);
  rpc RemoveRole(RemoveRoleRequest) returns (google.protobuf.Empty);
  rpc GetUserRoles(GetUserRolesRequest) returns (GetUserRolesResponse);

  // Permission checks
  rpc CheckPermission(CheckPermissionRequest) returns (CheckPermissionResponse);
}

// Authentication Messages
message LoginRequest {
  string email = 1;
  string password = 2;
}

message LoginResponse {
  string access_token = 1;
  string refresh_token = 2;
  User user = 3;
  int64 expires_in = 4;
}

message RegisterRequest {
  string email = 1;
  string password = 2;
  string first_name = 3;
  string last_name = 4;
  string phone = 5;
}

message RegisterResponse {
  string access_token = 1;
  string refresh_token = 2;
  User user = 3;
}

message RefreshTokenRequest {
  string refresh_token = 1;
}

message RefreshTokenResponse {
  string access_token = 1;
  string refresh_token = 2;
  int64 expires_in = 3;
}

message ValidateTokenRequest {
  string token = 1;
}

message ValidateTokenResponse {
  bool valid = 1;
  string user_id = 2;
  repeated string roles = 3;
  string error = 4;
}

message LogoutRequest {
  string user_id = 1;
  string token = 2;
}

// User CRUD Messages
message CreateUserRequest {
  string email = 1;
  string password = 2;
  string first_name = 3;
  string last_name = 4;
  string phone = 5;
  repeated string roles = 6;
}

message GetUserRequest {
  string user_id = 1;
}

message UpdateUserRequest {
  string user_id = 1;
  optional string email = 2;
  optional string first_name = 3;
  optional string last_name = 4;
  optional string phone = 5;
  optional bool is_active = 6;
}

message DeleteUserRequest {
  string user_id = 1;
}

message ListUsersRequest {
  int32 page = 1;
  int32 page_size = 2;
  string search = 3;
  optional string role = 4;
  optional bool is_active = 5;
}

message ListUsersResponse {
  repeated User users = 1;
  int32 total = 2;
  int32 page = 3;
  int32 page_size = 4;
}

message UserResponse {
  User user = 1;
}

// Profile Messages
message GetProfileRequest {
  string user_id = 1;
}

message UpdateProfileRequest {
  string user_id = 1;
  optional string first_name = 2;
  optional string last_name = 3;
  optional string phone = 4;
  optional string avatar_url = 5;
  optional string bio = 6;
}

message ProfileResponse {
  Profile profile = 1;
}

message ChangePasswordRequest {
  string user_id = 1;
  string old_password = 2;
  string new_password = 3;
}

// Role Messages
message AssignRoleRequest {
  string user_id = 1;
  string role = 2;
}

message RemoveRoleRequest {
  string user_id = 1;
  string role = 2;
}

message GetUserRolesRequest {
  string user_id = 1;
}

message GetUserRolesResponse {
  repeated string roles = 1;
}

message CheckPermissionRequest {
  string user_id = 1;
  string permission = 2;
}

message CheckPermissionResponse {
  bool has_permission = 1;
}

// Data Models
message User {
  string id = 1;
  string email = 2;
  string first_name = 3;
  string last_name = 4;
  string phone = 5;
  repeated string roles = 6;
  bool is_active = 7;
  google.protobuf.Timestamp created_at = 8;
  google.protobuf.Timestamp updated_at = 9;
}

message Profile {
  string user_id = 1;
  string first_name = 2;
  string last_name = 3;
  string email = 4;
  string phone = 5;
  string avatar_url = 6;
  string bio = 7;
  repeated string roles = 8;
  google.protobuf.Timestamp created_at = 9;
  google.protobuf.Timestamp updated_at = 10;
}
